### PR TITLE
feat: get message by id

### LINKED ISSUE #254: Download media later
Hello! I have a doubt. Is there a possibility to download a media afterwards? That is, by the message id can I retrieve it and download it?

### LINKED ISSUE #354: Ability to fetch a message by ID
Forgive me if the feature already exists and I just didn't look hard enough, but I'm pretty sure that it doesn't exist yet.
The feature should probably be somewhere inside the Chat structure, similar to how Client has a .getChatById method.

> Chat.getMessageById(messageID);

This would be useful for replying to or quoting any given message.

### LINKED ISSUE #656: Delete (revoke) a message per ID
Hello,

Is it possible to revoke a message for everyone per message id?

as I can see the message_revoke_everyone is limited only for the latest message only (last one)

looking forward.

### LINKED ISSUE #919: Bug downloadMedia()
### Bug description
The bug happens to me when I send the same file (GIF / MP4) 2 times.

Let me explain if I send x GIF once it will arrive correctly and I can download it, but if I later send the same GIF again and try to download it for some reason, it will return 'undefined',

I thought I was the one who programmed it wrong but I did a loop to test my idea when it sends a message Message.type = 'video' and exactly what I say happens.

### Reproduction steps
1. The client should receive a video-type message
2. The client must download the content with downloadMedia()
3. Upon successful download, download it again and it will return undefined.
...

### Expected behavior 
the media will download again.

### Relevant code
`                    setInterval( async () => {
                        let Video = await Message.downloadMedia();
                        console.log(Video);
                    }, 1000);
`

### Environment (please complete the following information):
**WhatsApp**
 - Account type [Standard / Business]: Standar
 - Device OS [iOS / Android]: Android
 - WhatsApp Web version [run `await client.getWWebVersion()`]: 2.2138.13

**Library**
 - Browser [Chrome / Chromium]: Chrome/Chromium
 - NodeJS version (`node -v`):  v16.7.0
 - npm or yarn version (`npm -v`): 7.20.3
 - whatsapp-web.js version: 1.14.1

**Other**
 - Operating system (the one running node) [Linux / MacOS / Windows]: Windows
 - Operating system version (ex. Windows 10):  Windows 11 and Windows 10 (Same result)

### Additional context
Screenshot loop code
Screenshot_1

Excuse me for my English, it is not my native language.

### LINKED ISSUE #979: Problem downloading media that has been sent for some time from a quotedMessage
### Bug description

**I'm using the multi device version**

I'm having problems when I try to download the media when someone replies to a message that has been sent some time ago.
For some reason when calling `downloadMedia()` the message becomes undefined.
This problem only happens with some old messages (hours ago).

### Reproduction steps
```
const quotedMessage = await message.getQuotedMessage()
console.log(quotedMessage) // Here the message is an object
const media = await quotedMessage.downloadMedia()
console.log(quotedMessage)  // Here the message is an undefined
console.log(media) // undefined too
```

### Expected behavior 
I expected only the media to return undefined if there was an error... But the message goes undefined

### Environment (please complete the following information):
**WhatsApp**
 - Account type [Standard / Business]: Standard
 - Device OS [iOS / Android]: Android
 - WhatsApp Web version [run `await client.getWWebVersion()`]: 2.2144.11

**Library**
 - Browser [Chrome / Chromium]: Chromium
 - NodeJS version (`node -v`):  v14.15.1
 - npm or yarn version (`npm -v`): 6.14.8
 - whatsapp-web.js version: #multidevice

**Other**
 - Operating system (the one running node) [Linux / MacOS / Windows]: Linux
 - Operating system version (ex. Windows 10):  debian:buster
### Additional context
I'm using the multi device version

### PR DESCRIPTION
**Pull request by pedroslopez**

### CODE REVIEW

File: src/Client.js [hunk 1]
```diff
@@ -610,6 +610,37 @@ class Client extends EventEmitter {
         return ContactFactory.create(this, contact);
     }
 
+    async getMessageById(messageId) {
+        const msg = await this.pupPage.evaluate(async messageId => {
+            let msg = window.Store.Msg.get(messageId);
+            if(msg) return window.WWebJS.getMessageModel(msg);
+
+            const params = messageId.split('_');
+            if(params.length !== 3) throw new Error('Invalid serialized message id specified');
```
Comment (tuyuribr) [inline src/Client.js:619]: We need to check param.lenght before doing Msg.get (To avoid errors)

Comment (PurpShell) [reply src/Client.js:619]: good call

**Comment by PurpShell** [comment]
did some tests

**Comment by tuyuribr** [comment]
For reference:
Starts failing on 20 hours (70971 seconds)

the "message" that got success was 18 days ago, and the type is: 'e2e_notification'

image

**Comment by PurpShell** [comment]
> For reference:
> Starts failing on 20 hours (70971 seconds)
> 
> the "message" that got success was 18 days ago, and the type is: 'e2e_notification'
> 
> messageId, IfGotTheMessage, timeSpent, timeDiference
> image

that's some extensive testing

**Comment by tuyuribr** [comment]
It's failing randomly, if someone notice a patern please help us (got messageId from msgstore.db `select timestamp,key_remote_jid,key_id,key_from_me,media_wa_type from messages order by timestamp desc limit 1000` )

[tests.csv](https://github.com/pedroslopez/whatsapp-web.js/files/7080113/tests.csv)

**Comment by akaa0** [comment]
When trying to get an old message, I get null.
And this message is really not deleted.

image

**Comment by PurpShell** [comment]
> When trying to get an old message, I get null.
> And this message is really not deleted.
> 
> image

Does this happen always? There is also a chance that the chat has been cleared etc.. Make sure you can search for it via the search bar...

**Comment by ReniDelonzek** [comment]
Any chance this works on MD?

**Comment by PurpShell** [comment]
> Any chance this works on MD?

Haha, no. This would only fetch messages stored in the storage of a linked device. No fetching from phone

**Comment by hanumanjiblog** [comment]
I need this feature please merge it as soon as possible

**Comment by shirser121** [comment]
There is a new way to fix the problem.
```async function (messageId) {
        const msg = await this.pupPage.evaluate(async messageId => {
            let msg = window.Store.Msg.get(messageId);
            if(msg) return window.WWebJS.getMessageModel(msg);
            const params = messageId.split('_');
            if(params.length !== 3) throw new Error('Invalid serialized message id specified');

            const chatId = params[1];
            const chat = window.Store.Chat.get(chatId);
            while (chat.msgs._models.find(msg => msg.id._serialized === messageId)) {
                await chat.loadEarlierMsgs();
            }
            msg = window.Store.Msg.get(messageId);
            if(msg) return window.WWebJS.getMessageModel(msg);
        }, messageId);

        if(msg) return new Message(this, msg);
        return null;
    }

**Review by shirser121 — CHANGES_REQUESTED** [review]
What about
```async function (messageId) {
        const msg = await this.pupPage.evaluate(async messageId => {
            let msg = window.Store.Msg.get(messageId);
            if(msg) return window.WWebJS.getMessageModel(msg);
            const params = messageId.split('_');
            if(params.length !== 3) throw new Error('Invalid serialized message id specified');

            const chatId = params[1];
            const chat = window.Store.Chat.get(chatId);
            while (chat.msgs._models.find(msg => msg.id._serialized === messageId)) {
                await chat.loadEarlierMsgs();
            }
            msg = window.Store.Msg.get(messageId);
            if(msg) return window.WWebJS.getMessageModel(msg);
        }, messageId);

        if(msg) return new Message(this, msg);
        return null;
    }

**Comment by shirser121** [comment]
> What about
> 
> ```
>         const msg = await this.pupPage.evaluate(async messageId => {
>             let msg = window.Store.Msg.get(messageId);
>             if(msg) return window.WWebJS.getMessageModel(msg);
>             const params = messageId.split('_');
>             if(params.length !== 3) throw new Error('Invalid serialized message id specified');
> 
>             const chatId = params[1];
>             const chat = window.Store.Chat.get(chatId);
>             while (chat.msgs._models.find(msg => msg.id._serialized === messageId)) {
>                 await chat.loadEarlierMsgs();
>             }
>             msg = window.Store.Msg.get(messageId);
>             if(msg) return window.WWebJS.getMessageModel(msg);
>         }, messageId);
> 
>         if(msg) return new Message(this, msg);
>         return null;
>     }
> ```

@pedroslopez This way seems working

**Comment by roigreenberg** [comment]
@pedroslopez is there any estimation for merging this PR?

**Comment by herduin** [comment]
@pedroslopez this OR is planned to merge? or what workaround you can recommend me?

**Review by codebossdev — APPROVED** [review]

**Comment by shirser121** [comment]
Closing after merge of #1634

**Pull request closed without merging**